rasa/utils/tensorflow/layers.py

```diff
@@ -1085,3 +1140,314 @@ def call(
+            tf.expand_dims(batch_labels_embed[:, 0, ...], axis=-2),
+            axis=1,
+            name="expand_pos_labels",
+        )
+
+        # Pick random examples from the batch
+        candidate_ids = self._get_candidate_indices(
+            batch_size=tf.shape(batch_inputs_embed)[0],
+            total_candidates=tf.shape(all_labels_embed)[0],
+        )
+
+        # Get the label embeddings and ids corresponding to candidate indices
+        candidate_labels_embed = self._get_candidate_values(
+            all_labels_embed, candidate_ids
+        )
+        candidate_labels_ids = self._get_candidate_values(all_labels_ids, candidate_ids)
+
+        # Determine how many distinct labels exist (highest label index)
+        max_label_id = tf.cast(tf.math.reduce_max(all_labels_ids), dtype=tf.int32)
+
+        # Convert the positive label ids to their one_hot representation.
+        # Note: -1 indices yield a zeros-only vector
+        batch_labels_one_hot = tf.one_hot(
+            tf.cast(tf.squeeze(batch_labels_ids, axis=-1), tf.int32),
+            max_label_id + 1,
+            axis=-1,
+        )  # (batch_size, max_num_labels_per_input, depth_needed)
+
+        # Collapse the extra dimension and convert to a multi-hot representation
+        # by aggregating all ones in the one-hot representation.
+        # Here tf.reduce_any is important and tf.reduce_sum
+        # cannot be used. Reason being that due to padding in label_ids,
+        # there can be more than one 1's at the 0'th index of the last dimension.
+        # Otherwise the loss function will be highly unstable.
+        batch_labels_multi_hot = tf.cast(
+            tf.math.reduce_any(tf.cast(batch_labels_one_hot, dtype=tf.bool), axis=-2),
+            tf.float32,
+        )  # (batch_size, depth_needed)
+
+        # Remove extra dimensions for gather
+        candidate_labels_ids = tf.squeeze(tf.squeeze(candidate_labels_ids, 1), -1)
+
+        # Collect the labels corresponding to negative
+        # label ids sampled for each data point of the batch earlier.
+        pos_neg_labels = tf.gather(
+            batch_labels_multi_hot,
+            tf.cast(candidate_labels_ids, tf.int32),
+            batch_dims=1,
+            name="gather_labels",
+        )
+
+        pos_neg_labels = tf.cast(pos_neg_labels, tf.float32)
+
+        return (
+            pos_inputs_embed,
+            pos_labels_embed,
+            candidate_labels_embed,
+            pos_neg_labels,
+        )
+
+    def _get_candidate_indices(
+        self, batch_size: tf.Tensor, total_candidates: tf.Tensor
+    ) -> tf.Tensor:
+        """Returns batch of random candidate indices
+        
+        Args:
+            batch_size: Size of zero'th dimension
+            total_candidates: Size of first dimension
+
+        Returns:
+            tf.int32 tensor of shape (target_size, total_candidates) with values 
+            in [0, self.num_neg).
+        """
```

Comment:
Do we need this as an extra function? Can't we directly call `layer_utils.random_indices` just as how `_get_negs` of `SingleLabelDotProductLoss` class does?